### PR TITLE
Make run-dependent MC configurable with cmsDriver options

### PR DESCRIPTION
**Pull request by christopheralanwest**
#### PR description:

For Run 3, it is intended to support the use of run-dependent MC using the scheme described in the [Time-Dependent MC Working group final recommendations](https://docs.google.com/document/d/1GLADYZP_ad1bPyrEBPpkRPT22Oru6fC_KKZmfgDgwZo/edit#). The proposed run assignment scheme uses a fixed mapping of lumiblock number to run number. Though this procedure is technically supported already by CMSSW, the lumiblock to run mapping required currently must be constructed manually from the run probability distribution, the number of lumiblocks in the sample and the events per lumiblock. This mapping, `firstLuminosityBlockForEachRun` is then specified individually for each workflow.

This PR adds two options for configuring this mapping from `cmsDriver.py` options. The first, `--runsAndWeightsForMCInteger`, allows for directly specifying a run probability distribution. In contrast to the Run 1 procedure, here the relative weighting of events per IOV is a rational number. As an example,

```
--runsAndWeightsForMCInteger "[315257, 315257, 316082]"
```

would result in two IOVs, one starting at `run = 315257` and the second starting at `run = 316082` with twice as many events in the IOV starting at `run = 315257`.

The options `--runsAndWeightsForMCInteger` and `--runsScenarioForMCInteger` are directly analogous to the options  `--runsAndWeightsForMC` and `--runsScenarioForMC` for the Run-1 run-dependent MC scheme. I decided to introduce new options rather than try to overload the existing options depending upon the input. I could of course change this behavior, if desired.

In order to properly construct the lumi-to-run mapping for the digi-premixing step, it is necessary to know the total number of lumiblocks in the sample. This is determined by using the existing `--relval` option used by the GEN-SIM step also for the digi-premixing step. In this PR, I am focusing on the relval workflows as it is desirable to use relvals to test run-dependent MC procedures prior to proposing changes needed for large productions.

To simplify the configuration of run-dependent MC, I've introduced a new process modifier `runDependent` to contain configuration options specific to run-dependent MC. For now, this amounts to using the setting

```
process.mixData.workers.ecal.timeDependent=True
```

in the DIGI-PREMIX step and forcing the run number to 1 in the HARVEST step.

Finally, the default settings for the run-dependent MC workflows have been changed so that the number of lumiblocks in the same evenly divides the number of IOVs. This is necessary to apply the IOV weighting with 9 evenly weighted IOVs, as used by recent tests of run-dependent MC.

#### PR validation:

This PR was tested by dumping the python files for all of the workflows `250202.172,250202.183,250206.182,250200.182,250202.182` with `edmConfigDump` and comparing the results before and after the changes.

For standard 2018 run-dependent workflows 250200.182, 250202.182 and 250206.182, corresponding to ZEE, TTbar and ZMM, the changes in each step are as follows:

GEN-SIM:
- Added `process.EcalLaserCorrectionServiceMC = cms.ESProducer("EcalLaserCorrectionServiceMC")`
- Modified number of events per lumiblock: `numberEventsInLuminosityBlock = cms.untracked.uint32(5)` -> `numberEventsInLuminosityBlock = cms.untracked.uint32(50)`

step 2:
- Modified lumiblock-to-run mapping. Changed from:

```
process.source.firstLuminosityBlockForEachRun = cms.untracked.VLuminosityBlockID(*[cms.LuminosityBlockID(x,y) for x,y in ((315257, 1), (316082, 222), (316720, 445), (317527, 668), (320917, 890), (321414, 1112), (321973, 1334), (322492, 1556), (324245, 1779))]) 
```

to 
```
process.source.firstLuminosityBlockForEachRun = cms.untracked.VLuminosityBlockID(*[cms.LuminosityBlockID(x,y) for x,y in ((315257, 1), (316082, 21), (316720, 41), (317527, 61), (320917, 81), (321414, 101), (321973, 121), (322492, 141), (324245, 161))])
```
step 3:
- No Changes

step 4:
- Added `process.EcalLaserCorrectionServiceMC = cms.ESProducer("EcalLaserCorrectionServiceMC")`

The additional ESProducer in the GEN-SIM step and step 4 is harmless as nothing in the GEN-SIM or HARVESTING step can consume this product.

For workflow 250202.183, which is a version of workflow 250202.182 that is intended to run multi-IOV MC jobs in IBs, the only changes are the additional `ESProducer` in the GEN-SIM step and step 4. No changes are needed for the lumiblock mapping in this case as the lumi-to-run mapping is configured differently for this workflow.

In the 2017 run-dependent workflow 250202.172, the additional `ESProducer` has been added to the GEN-SIM step, step 3 and step 4. There is no lumiblock-to-run splitting in this case.

In addition, I verified that the options `--runsAndWeightsForMCInteger [315257, 316082, 316720, 317527, 320917, 321414, 321973, 322492, 324245]` and `--runsScenarioForMCInteger Run2018_Equal_Lumi_Integer_Weights` result in identical configurations, as intended.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This PR is not a backport and there is currently no use case for a backport.

### CODE REVIEW

**Comment by christopheralanwest** [comment]
please test

**Comment by christopheralanwest** [comment]
test parameters
workflow = 250202.172,250202.183,250206.182,250200.182,250202.182

**Comment by christopheralanwest** [comment]
please test

**Comment by christopheralanwest** [comment]
I re-triggered the tests because I forgot to test the run-dependent workflows, which are the main focus of this PR. 

In the previous round of tests, there were unexpected comparison test failures in standard run-independent MC. However, the failing comparisons in the `h_ele_neutralHadronRelativeIso_mAOD_recomp` histograms are unrelated to this PR, and are described in issue #33554.

**Comment by christopheralanwest** [comment]
@amassiro Do you have any comments on the reorganization in this PR?

**Comment by amassiro** [comment]
> @amassiro Do you have any comments on the reorganization in this PR?

Very nice improvement in the user interface: the simple option "runsAndWeightsForMCInteger" is quite intuitive.

Question: are the changes in step 2 mentioned in the description automatic once the option "runsAndWeightsForMCInteger" is used? If yes, great!

I let @ecal-dpg reco experts comment further if needed.

I suppose we need a relval sample to see if the changes vs run number is still working (and maybe DQM plots, as will be discussed on the 10th of May)

**Comment by christopheralanwest** [comment]
> > @amassiro Do you have any comments on the reorganization in this PR?
> 
> Very nice improvement in the user interface: the simple option "runsAndWeightsForMCInteger" is quite intuitive.
> 
> Question: are the changes in step 2 mentioned in the description automatic once the option "runsAndWeightsForMCInteger" is used? If yes, great!
> 

No, because the `runDependent` process modifier and the `runsAndWeightsForMCInteger`/`runsScenarioForMCInteger` are independent. For example, the `runDependent` modifier is also used in the `HARVEST` step to force the run number to 1, as required by the DQM GUI to handle run-dependent MC: https://github.com/cms-sw/cmssw/pull/33555/files#diff-f268b897071f4e85943193524eb3c12f111179831007da097cea86f5036b1d94R22. Also, the functionality to explicitly configure the lumiblock-to-run mapping rather than using the `cmsDriver.py` options, as used by workflow 250202.183 at https://github.com/cms-sw/cmssw/pull/33555/files#diff-b0dce787f15ea8526a7b5585e797ba4b9e17ba829550c8283b5b5f97cba58b04R1814, still exists.

> I let @ecal-dpg reco experts comment further if needed.
> 

This should be @cms-sw/ecal-dpg-l2.

> I suppose we need a relval sample to see if the changes vs run number is still working (and maybe DQM plots, as will be discussed on the 10th of May)

Sure. I think a standard relval with 9k events would suffice for this purpose. 

This PR makes no changes to any of the DQM sequences used by the run-dependent relvals. If changes to the DQM/Validation sequences are required, that could be addressed in a separate PR.

**Comment by simonepigazzini** [comment]
Hi,

everything looks good to us (ECAL). As Andrea said we will probably need RelVals to duly check that different runs consumes different payloads and to do that we indeed need how to add the relevant plots to the DQM.

ciao,

simone

**Comment by christopheralanwest** [comment]
+alca

- Simplifies configuration of run-dependent MC with the Run-3 scheme
- Run-dependent relval workflows use a consistent set of configuration parameters by default
- As all changes are configuration-only, validated by using `edmConfigDump` on the python configuration for each step of each run-dependent relval workflow
- No unexpected changes in run-independent MC or data workflows
- Changes signed off by ECAL DPG. ECAL is currently the only subsystem that has any code (as opposed to conditions-only) specific to run-dependent MC

File: CalibCalorimetry/EcalLaserCorrection/python/ecalLaserCorrectionServiceMC_cfi.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import FWCore.ParameterSet.Config as cms
+
+EcalLaserCorrectionServiceMC = cms.ESProducer("EcalLaserCorrectionServiceMC")
```
Comment (silviodonato) [inline CalibCalorimetry/EcalLaserCorrection/python/ecalLaserCorrectionServiceMC_cfi.py:3]: The `_cfi` file should be left to be generated automatically with the fillDescriptions()
See https://cms-sw.github.io/cms_coding_rules.html and https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Converting_an_Existing_CFI_file

Comment (christopheralanwest) [reply CalibCalorimetry/EcalLaserCorrection/python/ecalLaserCorrectionServiceMC_cfi.py]: The latest commit (4147e8f) makes this change.

**Comment by francescobrivio** [comment]
test parameters
 - workflow = 250202.172,250202.183,250206.182,250200.182,250202.182

**Comment by francescobrivio** [comment]
@cmsbuild please test

**Comment by francescobrivio** [comment]
+alca

**Comment by silviodonato** [comment]
+operations

**Comment by silviodonato** [comment]
do you have any comments @cms-sw/pdmv-l2 @cms-sw/upgrade-l2 ?

**Comment by chayanit** [comment]
+1

**Comment by srimanob** [comment]
Hi  @christopheralanwest 

Sorry for my very late response.

For clarification,
- `--runsAndWeightsForMCInteger`, why you choose "Integer" in the name? I understand that you don't want to overwrite `--runsAndWeightsForMC` of Run-1, but I may misunderstand how the name is chosen.
- Why can we not use `era` to specify how the thing works?

Additional question, not related to the PR, Can we do Run-Dependent MC with classical mixing?

**Comment by christopheralanwest** [comment]
> For clarification,
> 
>     * `--runsAndWeightsForMCInteger`, why you choose "Integer" in the name? I understand that you don't want to overwrite `--runsAndWeightsForMC` of Run-1, but I may misunderstand how the name is chosen.
> 

In the Run-1 run-dependent MC IOV weighting method, the run assignment was performed per-event with the relative IOV weights assigned using floating point numbers representing the integrated luminosity of the IOV. In the Run-3 method, where a fixed lumiblock to run number mapping is used, the run weights are integers. I could have used `runsAndWeightsForMCIntegerWeights` and `runsScenarioForMCIntegerWeights` instead of `runsAndWeightsForMCInteger` and `runsScenarioForMCInteger`, respectively, but these option names seemed too long. Maybe the longer, more descriptive names are better?

>     * Why can we not use `era` to specify how the thing works?
> 
> 

PPD has a goal of eventually supporting multi-year run-dependent MC for Run 3. For this to be possible, we cannot have separate eras for the different years. Since only a single era-independent modification should be used for run-dependent MC, it is better to have a single process modifier rather than several eras that must be maintained identically.

> Additional question, not related to the PR, Can we do Run-Dependent MC with classical mixing?

In principle, I suppose it's possible. The challenge would be that the pileup events must be taken from the same IOV as the hard scatter event. I don't know that the code to do this matching (or, what's more, to do the matching efficiently) is available, though I believe that the mixing module does support IOV-dependent configuration. This is well beyond my area of expertise, however.

Is there a particular use case for run-dependent MC that requires classical mixing?

**Comment by srimanob** [comment]
> In the Run-1 run-dependent MC IOV weighting method, the run assignment was performed per-event with the relative IOV weights assigned using floating point numbers representing the integrated luminosity of the IOV. In the Run-3 method, where a fixed lumiblock to run number mapping is used, the run weights are integers. I could have used `runsAndWeightsForMCIntegerWeights` and `runsScenarioForMCIntegerWeights` instead of `runsAndWeightsForMCInteger` and `runsScenarioForMCInteger`, respectively, but these option names seemed too long. Maybe the longer, more descriptive names are better?
> 
The current one does not make sense much. However, if the conclusion that the shorter is better, I am OK with that.


> > ```
> > * Why can we not use `era` to specify how the thing works?
> > ```
> 
> PPD has a goal of eventually supporting multi-year run-dependent MC for Run 3. For this to be possible, we cannot have separate eras for the different years. Since only a single era-independent modification should be used for run-dependent MC, it is better to have a single process modifier rather than several eras that must be maintained identically.

Oh, I mean if `--era == Run1` then it picks up old customization, else (default) pick others. In this way, user don't need to choose `--runsAndWeightsXXX` because `era `can handle it properly. Am I missing something on this?

**Comment by makortel** [comment]
> Oh, I mean if `--era == Run1` then it picks up old customization, else (default) pick others. In this way, user don't need to choose `--runsAndWeightsXXX` because `era `can handle it properly. Am I missing something on this?

The "Run 1 way" is a technical implementation not tied to Run 1 data taking period (as far as I've understood). We could use the "Run 1 way" for Run 3 simulation, or "Run 3 way" for Run 1 simulation. Therefore the technical implementations should not be strictly tied to data taking periods.

(as a side note we don't have an Era for Run 1, the baseline "era-less" configuration should correspond to Run 1)

**Comment by christopheralanwest** [comment]
> > In the Run-1 run-dependent MC IOV weighting method, the run assignment was performed per-event with the relative IOV weights assigned using floating point numbers representing the integrated luminosity of the IOV. In the Run-3 method, where a fixed lumiblock to run number mapping is used, the run weights are integers. I could have used `runsAndWeightsForMCIntegerWeights` and `runsScenarioForMCIntegerWeights` instead of `runsAndWeightsForMCInteger` and `runsScenarioForMCInteger`, respectively, but these option names seemed too long. Maybe the longer, more descriptive names are better?
> 
> The current one does not make sense much. However, if the conclusion that the shorter is better, I am OK with that.
> 

I changed my mind regarding the option name and have modified the names in commit 8c537cb. A shorter command-line option name isn't of much help if it doesn't make sense.

> > > ```
> > > * Why can we not use `era` to specify how the thing works?
> > > ```
> > 
> > 
> > PPD has a goal of eventually supporting multi-year run-dependent MC for Run 3. For this to be possible, we cannot have separate eras for the different years. Since only a single era-independent modification should be used for run-dependent MC, it is better to have a single process modifier rather than several eras that must be maintained identically.
> 
> Oh, I mean if `--era == Run1` then it picks up old customization, else (default) pick others. In this way, user don't need to choose `--runsAndWeightsXXX` because `era `can handle it properly. Am I missing something on this?

I don't want to use the era mechanism to specify a customization because the choice of IOVs is still under discussion, even for 2018. I think it's better to have to manually specify a weighting scheme that is consistent with the global tag being used to avoid inconsistency between the IOVs in the IOV weighting and those used by the GT. Once an IOV weighting has been finalized for a particular year, it may be worthwhile to revisit this suggestion.

**Comment by srimanob** [comment]
test parameters:

- workflows = 250202.172,250202.183,250206.182,250200.182,250202.182

**Comment by srimanob** [comment]
Please test

**Comment by srimanob** [comment]
Hi @christopheralanwest @makortel 

Thanks for the clarification. I think this PR is a step to move on for the RunDependent relvals workflow. It makes a workflow configurable in cmsDriver. It should be good to go. However, we still have to

- figure out with PnR for the production mode, then ConfigBuilder.py will be updated again
- for multi-year, we may need to review configurations that are triggered by era and handle them properly. 
- Then we may try to merge with Run-1 config

**Comment by srimanob** [comment]
+Upgrade

@civanch @francescobrivio @chayanit Could you please sign again? The last commit is the configuration name changing. Thanks.

**Comment by chayanit** [comment]
+1

**Comment by francescobrivio** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**